src/py/flwr/server/driver/grpc_driver.py

```diff
@@ -221,24 +227,43 @@ def push_messages(self, messages: Iterable[Message]) -> Iterable[str]:
                 "list has `None` for those messages (the order is preserved as passed "
                 "to `push_messages`). This could be due to a malformed message.",
             )
-        return list(res.message_ids)
+        # Store message IDs
+        msg_ids = list(res.message_ids)
+        self._message_ids = msg_ids
+        return msg_ids
 
-    def pull_messages(self, message_ids: Iterable[str]) -> Iterable[Message]:
-        """Pull messages based on message IDs.
+    def pull_messages(
+        self, message_ids: Optional[Iterable[str]] = None
+    ) -> Iterable[Message]:
+        """Pull messages from the SuperLink based on message IDs.
 
         This method is used to collect messages from the SuperLink that correspond to a
-        set of given message IDs.
+        set of given message IDs. If no message IDs are provided, it defaults to the
+        stored message IDs.
         """
-        # Pull Messages
-        res: PullResMessagesResponse = self._stub.PullMessages(
-            PullResMessagesRequest(
-                message_ids=message_ids,
-                run_id=cast(Run, self._run).run_id,
-            )
+        # Raise an error if no message IDs are provided and none are stored
+        if not self._message_ids:
+            raise ValueError("No message IDs to pull. Call `push_messages` first.")
+
+        # Allow an override but default to the stored pending IDs
+        message_ids_to_pull = (
+            message_ids if message_ids is not None else self._message_ids
         )
-        # Convert Message from Protobuf representation
-        msgs = [message_from_proto(msg_proto) for msg_proto in res.messages_list]
-        return msgs
+
+        def iter_msg() -> Generator[Message, None, None]:
```

Comment:
We don't need a generator here, I think.
```suggestion
        def iter_msg() -> Iterator[Message]:
```